promql/functions.go

```diff
@@ -1662,6 +1662,110 @@ func funcHistogramQuantile(vectorVals []Vector, _ Matrix, args parser.Expression
 	return enh.Out, annos
 }
 
+func validateQuantile(q float64, arg parser.Expr) error {
+	if math.IsNaN(q) || q < 0 || q > 1 {
+		return annotations.NewInvalidQuantileWarning(q, arg.PositionRange())
+	}
+	return nil
+}
+
+// === histogram_quantiles(Vector parser.ValueTypeVector, label parser.ValueTypeString, q0 parser.ValueTypeScalar, qs parser.ValueTypeScalar...) (Vector, Annotations) ===
+func funcHistogramQuantiles(vectorVals []Vector, _ Matrix, args parser.Expressions, enh *EvalNodeHelper) (Vector, annotations.Annotations) {
+	var (
+		inVec         = vectorVals[0]
+		quantileLabel = args[1].(*parser.StringLiteral).Val
+		qs            = make([]float64, 0, len(vectorVals[2:]))
+
+		annos annotations.Annotations
+	)
+
+	for i := 2; i < len(vectorVals); i++ {
+		q := vectorVals[i][0].F
+
+		if err := validateQuantile(q, args[i]); err != nil {
+			annos.Add(err)
+		}
+
+		qs = append(qs, q)
+	}
+
+	annos.Merge(enh.resetHistograms(inVec, args[1]))
+
+	for _, q := range qs {
+		var qStr string
+
+		getOrCreateLblsWithQuantile := func(lbls labels.Labels) labels.Labels {
+			if enh.signatureToLabelsWithQuantile == nil {
+				enh.signatureToLabelsWithQuantile = make(map[string]map[float64]labels.Labels)
+			}
+
+			enh.lblBuf = lbls.Bytes(enh.lblBuf)
+			cachedLbls, ok := enh.signatureToLabelsWithQuantile[string(enh.lblBuf)]
+			if !ok {
+				enh.signatureToLabelsWithQuantile[string(enh.lblBuf)] = make(map[float64]labels.Labels, len(qs))
+			}
+
+			cachedLblsWithQuantile, ok := cachedLbls[q]
+			if !ok {
+				if qStr == "" {
+					qStr = fmt.Sprintf("%g", q)
+				}
+				cachedLblsWithQuantile = labels.NewBuilder(lbls).
+					Set(quantileLabel, qStr).
+					Labels()
+
+				enh.signatureToLabelsWithQuantile[string(enh.lblBuf)] = cachedLbls
```

Comment:
I believe this should be updating `cachedLbls[q]`?
```suggestion
				cachedLbls[q] = cachedLblsWithQuantile
```